staging/src/k8s.io/client-go/discovery/cached/memcache.go

```diff
@@ -117,35 +119,34 @@ func (d *memCacheClient) Fresh() bool {
 func (d *memCacheClient) Invalidate() {
 	d.lock.Lock()
 	defer d.lock.Unlock()
+	d.cacheValid = false
+}
 
+// refreshLocked refreshes the state of cache. The caller must hold d.lock for
+// writing.
+func (d *memCacheClient) refreshLocked() error {
 	// TODO: Could this multiplicative set of calls be replaced by a single call
 	// to ServerResources? If it's possible for more than one resulting
 	// APIResourceList to have the same GroupVersion, the lists would need merged.
 	gl, err := d.delegate.ServerGroups()
 	if err != nil || len(gl.Groups) == 0 {
-		utilruntime.HandleError(fmt.Errorf("couldn't get current server API group list; will keep using cached value. (%v)", err))
-		return
+		return fmt.Errorf("couldn't get current server API group list; will keep using cached value. (%v)", err)
 	}
 
 	rl := map[string]*metav1.APIResourceList{}
 	for _, g := range gl.Groups {
 		for _, v := range g.Versions {
 			r, err := d.delegate.ServerResourcesForGroupVersion(v.GroupVersion)
 			if err != nil || len(r.APIResources) == 0 {
-				utilruntime.HandleError(fmt.Errorf("couldn't get resource list for %v: %v", v.GroupVersion, err))
-				if cur, ok := d.groupToServerResources[v.GroupVersion]; ok {
-					// retain the existing list, if we had it.
-					r = cur
-				} else {
-					continue
-				}
+				return fmt.Errorf("couldn't get resource list for %v: %v", v.GroupVersion, err)
```

Comment:
It's not a good idea to fail the whole thing when a single resource fails. That typically means a misbehaving aggregated apiserver can cause all the other controllers to stop functioning. Whatever we do in code like this has to support partial failures.

I think reasonable behaviors are a combination of:
1. copy the thing out of the previous cache
2. leave an error marker in the cache, so you don't get the error until you try to access this groupversion
3. retry just this groupversion in the background

Reply:
I see the problem.

I think solution 1 is not a viable one, as this is against definition of 'Invalidate()' method in 'CachedDiscoveryInterface' which promises:

```
	// Invalidate enforces that no cached data is used in the future that is older than the current time.
	Invalidate()
```

2. The main reason for this PR is to make sure that transient errors doesn't cause permanent failure (right now if apiserver returns return 500 for some groupversion e.g. because it wasn't initialized yet it will cause errors forever).
3. I think I will try to implement something like that.

Reply:
I modified this to the state:
* store error in the cache
* if we are asked for an element that we know that exists, but there was a transient error, retry
* if the error is permanent (such element doesn't exist), we will simply return it.

We can add some rate-limiting for transient errors if needed.
Let me know what you think about this approach.